### PR TITLE
[GraphBolt] Overhaul `unique_and_compact` CPU.

### PR DESCRIPTION
**Pull request by mfbalin**
## Description
Using the new `std::atomic_ref` for portable compare and swap implementation. See #7614.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by mfbalin** [comment]
> Commit ID: [f1685ac](https://github.com/dmlc/dgl/commit/f1685ac2ed12bb38431b3ae058fceed417a9cd10)
> 
> Build ID: 2
> 
> Status: ❌ CI test failed in Stage [CPU Build].
> 
> Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-7613/2/2/logs/report.html)
> 
> Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-7613/2/2/logs/cireport.log)

@Rhett-Ying We need C++20 compliant compiler for this PR to go forward. As you can see, Windows supports parallel unique and compact implementation as well instead of the serial code.

File: graphbolt/CMakeLists.txt [hunk 1]
```diff
@@ -73,6 +73,8 @@ endif()
 add_library(${LIB_GRAPHBOLT_NAME} SHARED ${BOLT_SRC} ${BOLT_HEADERS})
 include_directories(BEFORE ${BOLT_DIR}
                            ${BOLT_HEADERS}
+                           # For CXX20 features.
```
Comment (Rhett-Ying) [inline graphbolt/CMakeLists.txt:76]: could you list what c++20 features are used for clarity?

**Review by Rhett-Ying — APPROVED** [review]

File: graphbolt/CMakeLists.txt (hunk 1 above)
Comment (mfbalin) [reply graphbolt/CMakeLists.txt:76]: Am I supposed to list them all in here each time I use something new?

Comment (mfbalin) [reply graphbolt/CMakeLists.txt:76]: This PR only uses `std::atomic_ref`

**Review by frozenbugs — APPROVED** [review]

File: graphbolt/CMakeLists.txt [hunk 2]
```diff
@@ -73,6 +73,9 @@ endif()
 add_library(${LIB_GRAPHBOLT_NAME} SHARED ${BOLT_SRC} ${BOLT_HEADERS})
 include_directories(BEFORE ${BOLT_DIR}
                            ${BOLT_HEADERS}
+                           # For CXX20 features:
+                           # `std::atomic_ref`, `std::counting_semaphore`
+                           "../third_party/cccl/libcudacxx/include"
```
Comment (frozenbugs) [inline graphbolt/CMakeLists.txt:78]: This add dependency to libcudacxx with CPU dgl, not sure about the consequences, let's keep an eye on it.

Comment (frozenbugs) [reply graphbolt/CMakeLists.txt:78]: Consider change this CMAKEList in a independent PR and see the result of regression before jumping to the merge of the real code.

Comment (mfbalin) [reply graphbolt/CMakeLists.txt:78]: There will be 0 effect on regression tests sue to the addition of this line in Cmake.

Comment (mfbalin) [reply graphbolt/CMakeLists.txt:78]: The stuff in it are not in the std namespace. It is in a completely new namespace.

Comment (mfbalin) [reply graphbolt/CMakeLists.txt:78]: If you don't use it, then it won't change anything. The namespace is `::cuda::std`, not `std`.

**Pull request merged**